### PR TITLE
fix(scripts): prioritize .specify over git for repo root detection

### LINKED ISSUE #1932: Bug: spec-kit uses parent git repo instead of .specify directory as root
## Description

When spec-kit is initialized in a subdirectory that doesn't have its own `.git`, but a parent directory does, spec-kit incorrectly uses the parent's git repository root instead of the directory containing `.specify`.

## Steps to Reproduce

1. Have a parent directory with git initialized: `/work/` (contains `.git`)
2. Create a subdirectory: `/work/my-project/`
3. Initialize spec-kit in the subdirectory: `cd /work/my-project && specify init`
4. Create a new feature: `specify new "My feature"`

## Expected Behavior

- Specs should be created in `/work/my-project/specs/`
- The `.specify` directory location should define the project root

## Actual Behavior

- Specs are created in `/work/specs/` (parent directory)
- `git rev-parse --show-toplevel` finds the parent's `.git` and uses that as root

## Root Cause

In `scripts/bash/common.sh` (lines 5-13), `get_repo_root()` prioritizes `git rev-parse --show-toplevel` over `.specify` directory detection:

```bash
get_repo_root() {
    if git rev-parse --show-toplevel >/dev/null 2>&1; then
        git rev-parse --show-toplevel    # <-- ALWAYS wins if ANY parent has .git
    else
        # Fallback only used when git completely fails
        ...
    fi
}
```

The same issue exists in:
- `scripts/bash/create-new-feature.sh` (lines 171-181)
- `scripts/powershell/common.ps1` (lines 4-16)
- `scripts/powershell/create-new-feature.ps1`

Ironically, there's a correct `find_repo_root()` function in `create-new-feature.sh` (lines 77-88) that checks for BOTH `.git` and `.specify`, but it's only used as a fallback when git fails completely.

## Proposed Fix

Invert the logic - `.specify` should be the primary marker for spec-kit root:

1. First search upward from CWD for `.specify` directory
2. If found, use that directory as the spec-kit root
3. Only use git as fallback if no `.specify` found

This ensures spec-kit respects its own initialization boundary rather than inheriting a parent git repo.

### PR DESCRIPTION
**Pull request by mbachorik**
## Summary

- Fixes bug where spec-kit uses parent git repo instead of `.specify` directory as root
- Adds `find_specify_root()` / `Find-SpecifyRoot` functions to search upward for `.specify` directory
- Updates `get_repo_root()` / `Get-RepoRoot` to check `.specify` first, then fall back to git
- Updates `has_git()` / `Test-HasGit` to verify git exists at the spec-kit root level (not parent)

## Problem

When spec-kit is initialized in a subdirectory that doesn't have its own `.git`, but a parent directory does, `git rev-parse --show-toplevel` finds the parent's git repo and uses that as root. This causes specs to be created in the wrong location.

**Example:**
- Parent `/work/` has `.git`
- Child `/work/my-project/` has `.specify` (spec-kit initialized here)
- Before: specs created in `/work/specs/` (wrong)
- After: specs created in `/work/my-project/specs/` (correct)

## Test plan

- [x] Verified bash syntax with `bash -n`
- [x] Verified PowerShell syntax with parser
- [x] Tested: child with `.specify`, parent with `.git` → returns child
- [x] Tested: project with both `.specify` and `.git` → returns project, has_git=true
- [x] Tested: only `.git`, no `.specify` → falls back to git correctly
- [x] Tested: deeply nested `.specify` → found correctly
- [x] Tested: `.specify` in parent, working in child → finds parent `.specify`
- [x] Tested: `.specify` as file (not directory) → correctly ignored

Fixes #1932

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: scripts/powershell/create-new-feature.ps1 [hunk 1]
```diff
@@ -135,27 +135,23 @@ function ConvertTo-CleanBranchName {
...
-    if ($LASTEXITCODE -eq 0) {
-        $hasGit = $true
-    } else {
-        throw "Git not available"
+# Use common.ps1 functions which prioritize .specify over git
+$repoRoot = Get-RepoRoot
+if (-not $repoRoot) {
+    # Fallback to local Find-RepositoryRoot if common.ps1 couldn't find it
+    $repoRoot = (Find-RepositoryRoot -StartDir $PSScriptRoot)
+    if (-not $repoRoot) {
```
[the remark this answers is not in the document]
Comment (mbachorik) [reply scripts/powershell/create-new-feature.ps1]: Fixed in 8a9ce68 - removed the dead fallback block since Get-RepoRoot always returns a path.

File: scripts/bash/create-new-feature.sh [hunk 2]
```diff
@@ -162,21 +162,24 @@ clean_branch_name() {
...
-else
+REPO_ROOT=$(get_repo_root)
+if [ -z "$REPO_ROOT" ]; then
+    # Fallback to local find_repo_root if common.sh couldn't find it
     REPO_ROOT="$(find_repo_root "$SCRIPT_DIR")"
     if [ -z "$REPO_ROOT" ]; then
         echo "Error: Could not determine repository root. Please run this script from within the repository." >&2
         exit 1
     fi
+fi
```
[the remark this answers is not in the document]
Comment (mbachorik) [reply scripts/bash/create-new-feature.sh]: Fixed in 8a9ce68 - removed the unreachable fallback block.

File: scripts/powershell/common.ps1 [hunk 3]
```diff
@@ -1,7 +1,34 @@
...
 # Common PowerShell functions analogous to common.sh
 
+# Find repository root by searching upward for .specify directory
+# This is the primary marker for spec-kit projects
+function Find-SpecifyRoot {
+    param([string]$StartDir = (Get-Location).Path)
+
+    $current = $StartDir
+    while ($true) {
+        if (Test-Path (Join-Path $current ".specify")) {
```
[the remark this answers is not in the document]
Comment (mbachorik) [reply scripts/powershell/common.ps1]: Fixed in 8a9ce68 - added `-PathType Container` to only match .specify directories.

File: scripts/powershell/common.ps1 [hunk 4]
```diff
@@ -58,13 +85,11 @@ function Get-CurrentBranch {
...
 function Test-HasGit {
-    try {
-        git rev-parse --show-toplevel 2>$null | Out-Null
-        return ($LASTEXITCODE -eq 0)
-    } catch {
-        return $false
-    }
+    $repoRoot = Get-RepoRoot
+    return (Test-Path (Join-Path $repoRoot ".git"))
 }
```
[the remark this answers is not in the document]
Comment (mbachorik) [reply scripts/powershell/common.ps1:116]: Fixed in 8a9ce68 - now checks `Get-Command git`, verifies .git exists (file or directory), and validates with `git rev-parse --is-inside-work-tree`.

File: scripts/bash/common.sh [hunk 5]
```diff
@@ -1,15 +1,39 @@
...
+find_specify_root() {
+    local dir="${1:-$(pwd)}"
+    while [ "$dir" != "/" ]; do
+        if [ -d "$dir/.specify" ]; then
+            echo "$dir"
+            return 0
+        fi
+        dir="$(dirname "$dir")"
+    done
+    return 1
```
[the remark this answers is not in the document]
Comment (mbachorik) [reply scripts/bash/common.sh:24]: Fixed in 8a9ce68 - path is now normalized to absolute via `cd "$dir" && pwd` at the start, and added `prev_dir` tracking to break if dirname stops changing.

File: scripts/bash/common.sh [hunk 6]
```diff
@@ -57,9 +81,11 @@ get_current_branch() {
...
 }
 
-# Check if we have git available
+# Check if we have git available at the spec-kit root level
+# Returns true only if the .specify root has its own .git directory
 has_git() {
-    git rev-parse --show-toplevel >/dev/null 2>&1
+    local repo_root=$(get_repo_root)
+    [ -d "$repo_root/.git" ]
 }
```
[the remark this answers is not in the document]
Comment (mbachorik) [reply scripts/bash/common.sh:104]: Fixed in 8a9ce68 - now checks `command -v git`, accepts .git as file or directory (`-e` instead of `-d`), and validates with `git -C "$repo_root" rev-parse --is-inside-work-tree`.

File: scripts/bash/common.sh [hunk 7]
```diff
@@ -1,15 +1,43 @@
...
+    local prev_dir=""
+    while [ "$dir" != "/" ] && [ "$dir" != "$prev_dir" ]; do
+        if [ -d "$dir/.specify" ]; then
+            echo "$dir"
+            return 0
+        fi
+        prev_dir="$dir"
+        dir="$(dirname "$dir")"
+    done
+    return 1
```
[the remark this answers is not in the document]
Comment (mbachorik) [reply scripts/bash/common.sh]: Fixed in 93cef2e - restructured the loop to check for .specify before the termination condition, so filesystem root is now properly checked.

File: scripts/bash/common.sh [hunk 8]
```diff
@@ -57,9 +89,17 @@ get_current_branch() {
...
 has_git() {
-    git rev-parse --show-toplevel >/dev/null 2>&1
+    local repo_root=$(get_repo_root)
+    # First check if git command is available
+    command -v git >/dev/null 2>&1 || return 1
+    # Check if .git exists (directory or file for worktrees/submodules)
+    [ -e "$repo_root/.git" ] || return 1
+    # Verify it's actually a valid git work tree
+    git -C "$repo_root" rev-parse --is-inside-work-tree >/dev/null 2>&1
 }
```
[the remark this answers is not in the document]
Comment (mbachorik) [reply scripts/bash/common.sh:104]: Fixed in 334df41 - `get_current_branch` now uses `has_git` check before calling git, and runs `git -C "$repo_root"` to scope to the spec-kit root. Same fix applied to PowerShell `Get-CurrentBranch`.

File: scripts/bash/create-new-feature.sh [hunk 9]
```diff
@@ -162,21 +162,16 @@ clean_branch_name() {
...
-# Resolve repository root. Prefer git information when available, but fall back
-# to searching for repository markers so the workflow still functions in repositories that
-# were initialised with --no-git.
+# Resolve repository root using common.sh functions which prioritize .specify over git
 SCRIPT_DIR="$(CDPATH="" cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)"
 source "$SCRIPT_DIR/common.sh"
 
-if git rev-parse --show-toplevel >/dev/null 2>&1; then
-    REPO_ROOT=$(git rev-parse --show-toplevel)
+REPO_ROOT=$(get_repo_root)
```
[the remark this answers is not in the document]
Comment (mbachorik) [reply scripts/bash/create-new-feature.sh:156]: Fixed in 334df41 - removed the unused `find_repo_root()` helper function.

File: scripts/powershell/create-new-feature.ps1 [hunk 10]
```diff
@@ -135,26 +135,14 @@ function ConvertTo-CleanBranchName {
...
-        $hasGit = $true
-    } else {
-        throw "Git not available"
-    }
-} catch {
-    $repoRoot = $fallbackRoot
-    $hasGit = $false
-}
+# Use common.ps1 functions which prioritize .specify over git
+$repoRoot = Get-RepoRoot
```
[the remark this answers is not in the document]
Comment (mbachorik) [reply scripts/powershell/create-new-feature.ps1:118]: Fixed in 334df41 - removed the unused `Find-RepositoryRoot` function.

File: scripts/bash/common.sh [hunk 11]
```diff
@@ -1,15 +1,47 @@
 #!/usr/bin/env bash
 # Common functions and variables for all scripts
 
-# Get repository root, with fallback for non-git repositories
+# Find repository root by searching upward for .specify directory
+# This is the primary marker for spec-kit projects
+find_specify_root() {
+    local dir="${1:-$(pwd)}"
+    # Normalize to absolute path to prevent infinite loop with relative paths
+    dir="$(cd "$dir" 2>/dev/null && pwd)" || return 1
```
[the remark this answers is not in the document]
Comment (mbachorik) [reply scripts/bash/common.sh]: Fixed in 43b3d55 - changed to `cd -- "$dir"` to handle paths starting with `-`.

File: scripts/bash/common.sh [hunk 12]
```diff
@@ -57,9 +90,17 @@ get_current_branch() {
...
 
-# Check if we have git available
+# Check if we have git available at the spec-kit root level
+# Returns true only if git is installed and the repo root is inside a git work tree
+# Handles both regular repos (.git directory) and worktrees/submodules (.git file)
 has_git() {
-    git rev-parse --show-toplevel >/dev/null 2>&1
+    local repo_root=$(get_repo_root)
+    # First check if git command is available
+    command -v git >/dev/null 2>&1 || return 1
```
[the remark this answers is not in the document]
Comment (mbachorik) [reply scripts/bash/common.sh]: Fixed in a2c7c5b - moved `command -v git` check before `get_repo_root` call.

File: scripts/powershell/common.ps1 [hunk 13]
```diff
@@ -1,7 +1,37 @@
...
+# This is the primary marker for spec-kit projects
+function Find-SpecifyRoot {
+    param([string]$StartDir = (Get-Location).Path)
+
+    # Normalize to absolute path to prevent issues with relative paths
+    $current = (Resolve-Path $StartDir -ErrorAction SilentlyContinue)?.Path
+    if (-not $current) { return $null }
+
+    while ($true) {
+        if (Test-Path -Path (Join-Path $current ".specify") -PathType Container) {
```
[the remark this answers is not in the document]
Comment (mbachorik) [reply scripts/powershell/common.ps1]: Fixed in 6df0d5f - changed to `Resolve-Path -LiteralPath` and `Test-Path -LiteralPath` to handle paths with wildcard characters.

File: scripts/powershell/common.ps1 [hunk 14]
```diff
@@ -58,9 +90,22 @@ function Get-CurrentBranch {
...
 
+# Check if we have git available at the spec-kit root level
+# Returns true only if git is installed and the repo root is inside a git work tree
+# Handles both regular repos (.git directory) and worktrees/submodules (.git file)
 function Test-HasGit {
+    $repoRoot = Get-RepoRoot
+    # First check if git command is available
+    if (-not (Get-Command git -ErrorAction SilentlyContinue)) {
+        return $false
+    }
```
[the remark this answers is not in the document]
Comment (mbachorik) [reply scripts/powershell/common.ps1]: Fixed in 6df0d5f - moved `Get-Command git` check before `Get-RepoRoot` call.

File: scripts/powershell/common.ps1 [hunk 15]
```diff
@@ -58,9 +91,22 @@ function Get-CurrentBranch {
...
+# Returns true only if git is installed and the repo root is inside a git work tree
+# Handles both regular repos (.git directory) and worktrees/submodules (.git file)
 function Test-HasGit {
+    # First check if git command is available (before calling Get-RepoRoot which may use git)
+    if (-not (Get-Command git -ErrorAction SilentlyContinue)) {
+        return $false
+    }
+    $repoRoot = Get-RepoRoot
+    # Check if .git exists (directory or file for worktrees/submodules)
+    if (-not (Test-Path (Join-Path $repoRoot ".git"))) {
```
[the remark this answers is not in the document]
Comment (mbachorik) [reply scripts/powershell/common.ps1]: Fixed in 028ac2e - added `-LiteralPath` to `Test-Path` for the `.git` check.

**Comment by mbachorik** [comment]
A small fix of annoying edge case.

**Comment by mbachorik** [comment]
@mnriem ready.

**Comment by mnriem** [comment]
Thank you!
